File: python/xl-client/dial_xl/total.py
from typing import Iterable

from dial_xl.events import notify_observer, ObservableObserver
from dial_xl.field import Field
from dial_xl.reader import _Reader


class Total(ObservableObserver):
    __before: str = ""
    __prefix: str = "total\n"
    __fields: list[Field | str]
    __after: str = ""
    __field_indices: dict[str, int]

    def __init__(self):
        self.__fields = []
        self.__field_indices = {}

    def to_dsl(self) -> str:
        """Converts total section to DSL format."""

        return (
            f"{self.__before}"
            f"{self.__prefix}"
            f"{''.join(field.to_dsl() for field in self.__fields)}"
            f"{self.__after}"
        )

    def get_field(self, name: str) -> Field:
        index = self._find_field(name)
        if index == -1:
            raise ValueError(f"Field '{name}' not found")

        return self.__fields[index]

    @notify_observer
    def add_field(self, field: Field):
        """Add a field to the table and invalidates compilation/computation results and sheet parsing errors"""
        if field.name in self.__field_indices:
            raise ValueError(f"Field '{field.name}' already exists")

        field._attach(self)
        self.__fields.append(field)
        self.__field_indices[field.name] = len(self.__fields) - 1

    @notify_observer
    def remove_field(self, name: str) -> Field:
        """Remove a field from the table and invalidates compilation/computation results and sheet parsing errors"""
        index = self._find_field(name)
        if index == -1:
            raise ValueError(f"Field '{name}' not found")

        field = self.__fields.pop(index)
        field._detach()
        self._update_field_indices()

        return field

    def _find_field(self, name: str) -> int:
        return self.__field_indices.get(name, -1)

    @property
    def field_names(self) -> Iterable[str]:
        """Enumerates field names"""
        return (field.name for field in self.fields)

    @property
    def fields(self) -> Iterable[Field]:
        """Enumerates fields"""
        return (field for field in self.__fields if isinstance(field, Field))

    @classmethod
    def _deserialize(cls, reader: _Reader) -> "Total":
        result = cls()
        result.__before = reader.next(lambda d: d["span"]["from"])
        fields = reader.entity.get("fields", [])
        result.__prefix = (
            reader.next(fields[0]["span"]["from"])
            if fields
            else reader.next(lambda d: d["span"]["to"])
        )
        for field_entity in fields:
            field_reader = reader.with_entity(field_entity)
            unparsed = field_reader.next_unparsed(lambda d: d["span"]["from"])
            if unparsed:
                result.__fields.append(unparsed + field_reader.till_linebreak())
            field = Field._deserialize(field_reader)
            result.__fields.append(field)
            field._attach(result)
            reader.position = field_reader.position

        result.__after = reader.before_next()
        result._update_field_indices()

        return result

    def _update_field_indices(self):
        self.__field_indices = {
            field.name: index
            for index, field in enumerate(self.__fields)
            if isinstance(field, Field)
        }

File: frontend/libs/parser/src/lib/EditDslApi/Total.ts
import { lineBreak } from '../parser';
import { Field } from './Field';
import { notifyObserver, ObservableObserver, Reader } from './utils';

export class Total extends ObservableObserver {
  /**
   * Text that appears before the "total" keyword (often whitespace or comments).
   */
  private _before = '';

  /**
   * The prefix for this section, by default "total" plus a line break.
   */
  private _prefix = 'total' + lineBreak;

  /**
   * An array storing either raw text strings (e.g., unparsed snippets)
   * or actual {@link Field} objects in the order they appear in the DSL.
   */
  private _fields: Array<Field | string> = [];

  /**
   * Text that appears after all fields, often leftover content (whitespace or line breaks).
   */
  private _after = '';

  /**
   * A mapping of field names to their index in `_fields`, allowing quick lookups.
   */
  private _fieldIndices: Record<string, number> = {};

  /**
   * Constructs a new `Total` instance.
   */
  constructor() {
    super();
  }

  /**
   * Converts this total section (including prefix, fields, and trailing text)
   * back into a DSL string.
   *
   * @returns A DSL-formatted string representing this total section.
   */
  public toDSL(): string {
    return (
      this._before +
      this._prefix +
      this._fields
        .map((f) => (typeof f === 'string' ? f : f.toDSL()))
        .join('') +
      this._after
    );
  }

  /**
   * Retrieves a Field by its name.
   *
   * @param name - The name of the field to look up.
   * @returns The matching `Field` object.
   * @throws {Error} If the field is not found.
   */
  public getField(name: string): Field {
    const index = this.findField(name);
    if (index === -1) {
      throw new Error(`Field '${name}' not found`);
    }

    return this._fields[index] as Field;
  }

  /**
   * Adds a new field to the total section, triggering observer notifications.
   *
   * @param field - The {@link Field} object to add.
   * @throws {Error} If a field with the same name already exists.
   */
  @notifyObserver()
  public addField(field: Field): void {
    if (field.name in this._fieldIndices) {
      throw new Error(`Field '${field.name}' already exists`);
    }
    field.attach(this);
    this._fields.push(field);
    this._fieldIndices[field.name] = this._fields.length - 1;
  }

  /**
   * Removes and returns a field by name, triggering observer notifications.
   *
   * @param name - The name of the field to remove.
   * @returns The removed `Field` object.
   * @throws {Error} If the named field does not exist.
   */
  @notifyObserver()
  public removeField(name: string): Field {
    const index = this.findField(name);
    if (index === -1) {
      throw new Error(`Field '${name}' not found`);
    }
    const field = this._fields.splice(index, 1)[0] as Field;
    field.detach();
    this.updateFieldIndices();

    return field;
  }

  /**
   * Returns the index of a field in `_fields` by name, or -1 if not found.
   *
   * @param name - The field name to look for.
   */
  private findField(name: string): number {
    return this._fieldIndices[name] ?? -1;
  }

  /**
   * Enumerates the names of all fields in this total section.
   */
  public get fieldNames(): Iterable<string> {
    return function* (this: Total) {
      for (const f of this._fields) {
        if (f instanceof Field) {
          yield f.name;
        }
      }
    }.call(this);
  }

  /**
   * Enumerates all Field objects in this total section.
   */
  public get fields(): Iterable<Field> {
    return function* (this: Total) {
      for (const f of this._fields) {
        if (f instanceof Field) {
          yield f;
        }
      }
    }.call(this);
  }

  /**
   * Deserializes a `Total` object from the given Reader,
   * reading the prefix, fields (and any unparsed snippets),
   * leftover text, and rebuilding field indices.
   *
   * @param reader - A `Reader` positioned at a total section in the DSL.
   * @returns A fully populated `Total` instance.
   */
  public static deserialize(reader: Reader): Total {
    const result = new Total();

    // `_before`
    result._before = reader.next((d) => d.span.from);

    // read fields
    const fieldsData = reader.entity?.fields ?? [];
    if (fieldsData.length > 0) {
      // prefix ends where first field starts
      result._prefix = reader.next(fieldsData[0].span.from);
    } else {
      // no fields, read prefix up to final
      result._prefix = reader.next((d) => d.span.to);
    }

    // parse each field
    for (const fieldEntity of fieldsData) {
      const fieldReader = reader.withEntity(fieldEntity);

      // Possibly read unparsed snippet before the field
      const unparsed = fieldReader.nextUnparsed((d) => d.span.from);
      if (unparsed) {
        // e.g. store raw text + linebreak
        result._fields.push(unparsed + fieldReader.tillLinebreak());
      }

      // now parse the actual Field object
      const field = Field.deserialize(fieldReader);
      result._fields.push(field);
      field.attach(result);

      reader.position = fieldReader.position;
    }

    // leftover text
    result._after = reader.beforeNext();

    // rebuild indices
    result.updateFieldIndices();

    return result;
  }

  /**
   * Recomputes `_fieldIndices` by enumerating all actual fields in `_fields`.
   */
  private updateFieldIndices(): void {
    this._fieldIndices = {};
    this._fields.forEach((entry, idx) => {
      if (entry instanceof Field) {
        this._fieldIndices[entry.name] = idx;
      }
    });
  }
}
